### PR TITLE
Emit a warning on non-hashable static arguments in jax.jit.

### PR DESCRIPTION
**Pull request by jblespiau**
The message looks like, e.g.:

Static argument (index 1) of type <class 'numpy.ndarray'> for function f is non-hashable. As this can lead to unexpected cache-misses, it will raise an error in a near future.

### CODE REVIEW

**Comment by jblespiau** [comment]
Fixed mypy (by ignoring it).

File: jax/api.py [hunk 1]
```diff
@@ -130,11 +131,12 @@ def jit(fun: Callable[..., T],
       arguments to treat as static (compile-time constant). Operations that only
       depend on static arguments will be constant-folded in Python (during
       tracing), and so the corresponding argument values can be any Python
-      object. Calling the jitted function with different values for these
-      constants will trigger recompilation. If the jitted function is called
-      with fewer positional arguments than indicated by ``static_argnums`` then
-      an error is raised. Arguments that are not arrays or containers thereof
-      must be marked as static. Defaults to ().
+      object. Static arguments should be hashable and comparable. Calling the
```
Comment (mattjj) [inline jax/api.py:134]: I think ["comparable" entails more than just equality comparison](https://docs.python.org/2.0/ref/comparisons.html). Instead, [just saying hashable entails both `__hash__` and `__eq__`](https://docs.python.org/3.1/glossary.html#term-hashable). So I suggest we either just say "Static arguments should be hashable." or else just inline the standard Python definition for explicitness:

```suggestion
      object. Static arguments should be hashable, meaning both ``__hash__`` and ``__eq__` are implemented. Calling the
```

File: jax/api_util.py [hunk 2]
```diff
@@ -74,16 +75,78 @@ def flatten_fun_nokwargs2(in_tree, *args_flat):
...
-  fixed_args = tuple([unit if i in dyn_argnums else wrap_hashably(arg)
-                      for i, arg in enumerate(args)])
+  fixed_args = tuple(
+      unit if i in dyn_argnums else wrap_hashably(arg)
+      for i, arg in enumerate(args))
   dyn_args = tuple(args[i] for i in dyn_argnums)
   return _argnums_partial(f, dyn_argnums, fixed_args), dyn_args
 
+
+def argnums_partial_except(f,
```
Comment (mattjj) [inline jax/api_util.py:105]:
```suggestion
def argnums_partial_except(f: lu.WrappedFun,
```

File: jax/api_util.py [hunk 3]
```diff
@@ -74,16 +75,78 @@ def flatten_fun_nokwargs2(in_tree, *args_flat):
...
+                           static_argnums: Tuple[int, ...],
+                           args,
```
Comment (mattjj) [inline jax/api_util.py:107]:
```suggestion
                           args: Tuple[Any],
```

File: jax/api_util.py [hunk 4]
```diff
@@ -74,16 +75,78 @@ def flatten_fun_nokwargs2(in_tree, *args_flat):
...
+                           raise_error=False):
```
Comment (mattjj) [inline jax/api_util.py:108]:
```suggestion
                           raise_error: bool = False) -> Tuple[lu.WrappedFun, Tuple[Any]]:
```

File: jax/api_util.py [hunk 5]
```diff
@@ -74,16 +75,78 @@ def flatten_fun_nokwargs2(in_tree, *args_flat):
...
+
+  Args:
+    f: The function being wrapped.
+    static_argnums: The indices of the static arguments.
+    args: The positional arguments.
+    raise_error: If `True`, an error will be raised instead of a warning.
+
+  Returns:
+    A tuple of:
+      The function wrapped into a partial with static-arguments wrapped.
```
Comment (mattjj) [inline jax/api_util.py:119]: It's not accurate to say the function is wrapped into a partial, because what's actually being returned is a lu.WrappedFun instance.

I think this docstring isn't saying much more than the type annotations above. We could either
1. adapt the language so that it's precise (like the type annotations)
2. delete most of the docstring.
I suggest the latter.

How about a one-line docstring that just says

```python
"""Version of ``argnums_partial`` that checks hashability of static_argnums."""
```

File: jax/api_util.py [hunk 6]
```diff
@@ -74,16 +75,78 @@ def flatten_fun_nokwargs2(in_tree, *args_flat):
...
+
+  Args:
+    f: The function being wrapped.
+    dync_arguments: The indices of the dynamic arguments (i.e. non-static
+      positial arguments).
+    args: The positional arguments.
+
+  Returns:
+    A tuple of:
+      The function wrapped into a partial with the non dynamic arguments
```
Comment (mattjj) [inline jax/api_util.py:90]: This docstring isn't quite accurate (first element of the return value isn't a partial, it's an lu.WrappedFun), and moreover I don't think this docstring conveys more than type annotations.

Can you revert the changes to `argnums_partial`, both in the code and in the docstring? IIUC it doesn't need to change in this PR.

File: jax/api_util.py (hunk 4 above)
Comment (mattjj) [inline jax/api_util.py:108]: Can we remove the `raise_error` argument? I don't think it's used by any callers, and having options on a function is kind of an anti-pattern. Instead, we can just have the version of the function body that prints a warning for now, and in the future edit it to raise an error. (If we want to make it switchable later, we could have a flag rather than an argument to an internal-only function.)

**Review by mattjj — CHANGES_REQUESTED** [review]
Thanks for this! I have some changes to suggest.

File: jax/api.py (hunk 1 above)
Comment (jblespiau) [reply jax/api.py:134]: Done (manually, copying the suggested edits

File: jax/api_util.py (hunk 6 above)
Comment (jblespiau) [reply jax/api_util.py:90]: Done.

File: jax/api_util.py (hunk 2 above)
Comment (jblespiau) [reply jax/api_util.py:105]: Done.

File: jax/api_util.py (hunk 4 above)
Comment (jblespiau) [reply jax/api_util.py:108]: Done.

Comment (jblespiau) [reply jax/api_util.py:108]: Done.

File: jax/api_util.py (hunk 5 above)
Comment (jblespiau) [reply jax/api_util.py:119]: Done.

**Comment by jblespiau** [comment]
mypy was unhappy with the returned value annotation, so i removed them:

: error: Incompatible return value type (got "Tuple[Any, Tuple[Any, ...]]", expected "Tuple[WrappedFun, Tuple[Any]]")

**Review by mattjj — APPROVED** [review]
LGTM!

**Comment by mattjj** [comment]
Thanks for your patience, and your hard work!

**Comment by tomweingarten** [comment]
A number of the examples pass dicts as static_args to create_model. Should the examples be updated so they don't throw these errors?

For example:
https://github.com/google/flax/blob/master/examples/wmt/train.py#L257

**Comment by jekbradbury** [comment]
Yes, they should! Can you open an issue on the Flax repo for that? (Unless there are also instances of this in the JAX repo.)
